evm/p2p/auth.py

```diff
@@ -0,0 +1,284 @@
+
+    @asyncio.coroutine
+    def connect(self):
+        reader, writer = yield from asyncio.open_connection(
+            host=self.remote.address.ip, port=self.remote.address.tcp_port)
+        return reader, writer
+
+    def derive_secrets(self, initiator_nonce, responder_nonce,
+                       remote_ephemeral_pubkey, auth_init, auth_ack):
+        """Derive base secrets from ephemeral key agreement."""
+        # ecdhe-shared-secret = ecdh.agree(ephemeral-privkey, remote-ephemeral-pubk)
+        ecdhe_shared_secret = ecies.ecdh_agree(
+            self.ephemeral_privkey, remote_ephemeral_pubkey)
+
+        # shared-secret = sha3(ecdhe-shared-secret || sha3(nonce || initiator-nonce))
+        shared_secret = keccak(
+            ecdhe_shared_secret + keccak(responder_nonce + initiator_nonce))
+
+        # aes-secret = sha3(ecdhe-shared-secret || shared-secret)
+        aes_secret = keccak(ecdhe_shared_secret + shared_secret)
+
+        # mac-secret = sha3(ecdhe-shared-secret || aes-secret)
+        mac_secret = keccak(ecdhe_shared_secret + aes_secret)
+
+        # setup sha3 instances for the MACs
+        # egress-mac = sha3.update(mac-secret ^ recipient-nonce || auth-sent-init)
+        mac1 = keccak_updateable(sxor(mac_secret, responder_nonce) + auth_init)
+        # ingress-mac = sha3.update(mac-secret ^ initiator-nonce || auth-recvd-ack)
+        mac2 = keccak_updateable(sxor(mac_secret, initiator_nonce) + auth_ack)
+
+        if self._is_initiator:
+            egress_mac, ingress_mac = mac1, mac2
+        else:
+            egress_mac, ingress_mac = mac2, mac1
+
+        return aes_secret, mac_secret, egress_mac, ingress_mac
+
+
+class HandshakeInitiator(HandshakeBase):
+    _is_initiator = True
+
+    def encrypt_auth_message(self, auth_message):
+        auth_init = ecies.encrypt(auth_message, self.remote.pubkey)
+        return auth_init
+
+    def create_auth_message(self, nonce):
+        ecdh_shared_secret = ecies.ecdh_agree(self.privkey, self.remote.pubkey)
+        secret_xor_nonce = sxor(ecdh_shared_secret, nonce)
+
+        # S(ephemeral-privk, ecdh-shared-secret ^ nonce)
+        # FIXME: Should use ecdsa_sign(), but can't because it hashes the msg before signing.
+        S = encode_signature(
+            *get_ecc_backend().ecdsa_raw_sign(
+                secret_xor_nonce,
+                self.ephemeral_privkey)
+        )
+
+        # S || H(ephemeral-pubk) || pubk || nonce || 0x0
+        flag = 0x0
+        auth_message = S + keccak(self.ephemeral_pubkey) + self.pubkey + \
+            nonce + ascii_chr(flag)
+        return auth_message
+
+    def decode_auth_ack_message(self, ciphertext):
+        try:
+            eph_pubkey, nonce, version = decode_ack_plain(ciphertext, self.privkey)
+        except AuthenticationError:
+            eph_pubkey, nonce, version = decode_ack_eip8(ciphertext, self.privkey)
+            self.got_eip8_auth = True
+        auth_ack = ciphertext
```

Comment:
this assignment struck me as strange.

Reply:
Good catch; it makes no sense for this method to do this and then return it afterwards, as this is the very argument passed into the method. Fixed so it no longer returns auth_ack and dropped the assignment